### DISCUSSION TITLE
94. 인덱스를 사용하면 좋은 경우와 사용을 피해야 하는 경우는 어떤 것이 있나요?

### DISCUSSION THREAD

**Discussion by jongmee** [jewelry-box/cs-interview#98] [General]
인덱스를 사용하면 좋은 경우와 사용을 피해야 하는 경우는 어떤 것이 있나요?

**Comment by linirini** [comment]
### 인덱스를 사용하면 좋은 경우
#### 1. 데이터가 많은 테이블
인덱스는 대량의 데이터에서 특정 레코드를 빠르게 찾을 수 있다.
#### 2. WHERE 절에서 자주 사용되는 칼럼
쿼리에서 자주 조건으로 사용되는 칼럼에 인덱스를 생성하면 성능이 향상될 수 있다.
#### 3. JOIN 또는 ORDER BY에서 사용되는 칼럼
쿼리가 여러 테이블을 조인하거나 결과를 정렬해야 할 때 인덱스를 사용하면 성능이 향상될 수 있다.
#### 4. 카디널리티가 높은 칼럼
인덱스가 걸린 칼럼이 고유한 값을 가질수록 많은 행이 필터링 되므로 성능이 향상될 수 있다.
### 인덱스 사용을 피해야 하는 경우
#### 1. 데이터의 변경이 잦은 경우
인덱스는 수정 작업 시 인덱스 재정렬을 위한 추가적인 I/O가 발생한다. 따라서, 변동이 많은 테이블에 너무 많은 인덱스를 설정하면 성능 저하를 초래할 수 있다.
#### 2. 작은 테이블
데이터가 적은 테이블에 대해서는 인덱스를 생성해도 큰 성능 상 이점을 제공하지 못할 수 있다.
#### 3. 인덱스 선택도가 낮을 때
인덱스를 통해 많은 행이 필터링 되지 않는다면, 전체 테이블 스캔과 큰 차이가 없을 수 있다.
#### 4. 카디널리티가 낮을 때
인덱스가 많은 동일한 값을 가진 열에 대해서는 큰 성능 상 이점을 제공하지 못할 수 있다.
